### DISCUSSION TITLE
Example Kubernetes (K8s) Groups configuration to Teleport Roles

### DISCUSSION THREAD

**Discussion by stevenGravy** [gravitational/teleport#8581] [Show and tell]
Here is an example for adding K8s groups that can be used in roles or user traits to define access for users accessing a K8s cluster through Teleport.

This example shows two groups, one that can list and view logs for pods , and another that can also open exec sessions in the example-dev namespace.

Prerequisite: Kubernetes cluster with RBAC enabled (note that some like Microk8s you have to specifically enable this). Prior to running this you should have deployed Teleport in a Kubernetes cluster or added Teleport as a gateway to that cluster (See https://goteleport.com/teleport/docs/kubernetes-ssh/). Either through Teleport or direct to the cluster you should have the ability to define namespaces, create pods, roles and bindings.

### Create a Namespace and a nginx pod to test access
```bash
$ kubectl create ns example-dev
namespace/example-dev created
$ kubectl run nginx --image=nginx -n example-dev
pod/nginx created
$ kubectl get po -n example-dev
NAME    READY   STATUS    RESTARTS   AGE
nginx   1/1     Running   0          14s
```

### Create roles and bindings for accessing pods within the example-dev namespace
```
$ cat > exampleroles.yaml <<EOF
apiVersion: rbac.authorization.k8s.io/v1
kind: Role
metadata:
  name: dev-exec
  namespace: example-dev
rules:
- apiGroups:
  - ""
  resources:
  - pods
  - pods/log
  verbs:
  - get
  - list
  - watch
- apiGroups:
  - ""
  resources:
  - pods/exec
  verbs:
  - create
---
apiVersion: rbac.authorization.k8s.io/v1
kind: Role
metadata:
  name: dev-logs
  namespace: example-dev
rules:
- apiGroups:
  - ""
  resources:
  - pods
  - pods/log
  verbs:
  - get
  - list
  - watch
EOF
```

### Insert roles definition
```bash
$ kubectl create -f exampleroles.yaml 
role.rbac.authorization.k8s.io/dev-exec created
role.rbac.authorization.k8s.io/dev-logs created
```

### Create rolebindings configuration
```bash
$ cat > examplebindings.yaml <<EOF
apiVersion: rbac.authorization.k8s.io/v1
kind: RoleBinding
metadata:
  name: dev-exec-binding
  namespace: example-dev
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: Role
  name: dev-exec
subjects:
- apiGroup: rbac.authorization.k8s.io
  kind: Group
  name: developer-exec
---
apiVersion: rbac.authorization.k8s.io/v1
kind: RoleBinding
metadata:
  name: dev-logs-binding
  namespace: example-dev
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: Role
  name: dev-logs
subjects:
- apiGroup: rbac.authorization.k8s.io
  kind: Group
  name: developer-logs
EOF
Insert roles bindings
$ kubectl create -f examplebindings.yaml 
rolebinding.rbac.authorization.k8s.io/dev-exec-binding created
rolebinding.rbac.authorization.k8s.io/dev-logs-binding created
```

### Assign K8s group
Now we can assign users or roles with the `developer-exec` or `developer-logs` group. The Teleport k8s guide has further examples.

For roles you can assign within the `kubernetes_groups`:

```bash
cat > developer-exec.yaml <<EOF
kind: role
version: v4
metadata:
  name: developer-exec
spec:
  allow:
    kubernetes_groups: ["developer-exec"]
    kubernetes_labels:
       '*': '*'
EOF
```

### Insert role
```bash
tctl create -f  developer-exec.yaml 
```
Now assign a user that role ( ex: tctl users add --roles=developer-exec joe) or assign it in the auth connector settings.

### Login and use K8s access to list and create a exec session within pods
Note that other attempts to access different namespaces is denied.

$ tsh status
> Profile URL:        https://teleport.example.com:3080
  Logged in as:       joe
  Cluster:            teleport
  Roles:              developer-exec
  Logins:             -teleport-nologin-31c303e8-0065-4766-996e-1cf9b5660bc9
  Kubernetes:         enabled
  Kubernetes cluster: "teelport"
  Kubernetes groups:  developer-exec
  Valid until:        2020-12-09 05:38:18 -0500 EST [valid for 11h59m0s]
  Extensions:         permit-port-forwarding, permit-pty

```bash
$ kubectl get po -n example-dev
NAME    READY   STATUS    RESTARTS   AGE
nginx   1/1     Running   0          42m
```bash

Other namespaces will be denied:
```bash
$ kubectl get po -n example-devother
Error from server (Forbidden): pods is forbidden: User "joe" cannot list resource "pods" in API group "" in the namespace "example-devother"
```

You can exec a session
```bash
$ kubectl exec -n example-dev nginx -- bash
root@nginx:/# df
Filesystem     1K-blocks    Used Available Use% Mounted on
overlay         98868448 3633560  95218504   4% /
tmpfs              65536       0     65536   0% /dev
tmpfs            2015812       0   2015812   0% /sys/fs/cgroup
/dev/sda1       98868448 3633560  95218504   4% /etc/hosts
shm                65536       0     65536   0% /dev/shm
tmpfs            2015812      12   2015800   1% /run/secrets/kubernetes.io/serviceaccount
tmpfs            2015812       0   2015812   0% /proc/acpi
tmpfs            2015812       0   2015812   0% /proc/scsi
tmpfs            2015812       0   2015812   0% /sys/firmware
root@nginx:/# exit
```

The exec session will list as a playable session now by default.

**Comment by WilliamLoy** [comment]
Very concise. Thank you
